serialization/include/rp_ser.h

```diff
@@ -0,0 +1,361 @@
+#include <rp_common.h>
+#include <rp_trans.h>
+
+/**
+ * @file
+ * @defgroup rp_ser Remote procedures serialization core
+ * @{
+ * @brief Remote procedures serialization core API.
+ */
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**@brief Serialization packet type.*/
+enum rp_ser_packet_type {
+	/** Serialization command packet. */
+	RP_SER_PACKET_TYPE_CMD          = 0x01,
+
+	/** Serialization event packet. */
+	RP_SER_PACKET_TYPE_EVENT,
+
+	/** Serialization command response packet. */
+	RP_SER_PACKET_TYPE_RSP,
+
+	/** Serialization transport reserved packet. */
+	RP_SER_PACKET_TRANSPORT_RESERVED = 128,
+
+	/** Serialization upper bound. */
+	RP_SER_PACKER_TYPE_MAX           = 255
+};
+
+/**@brief Command response handler type. */
+typedef rp_err_t (*cmd_rsp_handler_t)(CborValue *it);
+
+/**@brief Command decoder type. */
+typedef rp_err_t (*cmd_handler_t)(CborValue *it);
+
+/**@brief Event decoder type. */
+typedef rp_err_t (*evt_handler_t)(uint8_t evt, CborValue *it);
+
+/**@brief Encoder data structure. */
+struct rp_ser_encoder {
+	/** Main encoder. */
+	CborEncoder encoder;
+
+	/** Encoder container like array or map */
+	CborEncoder *container;
+
+	/** Encoder buffer. */
+	uint8_t *buf;
+
+	/** Encoder buffer size. */
+	size_t buf_size;
+
+	/** Packet size after encoding. */
+	size_t packet_size;
+};
+
+/**@brief Command decoder structure. */
+struct rp_ser_cmd {
+	/** Command code. */
+	uint8_t cmd;
+
+	/** Command decoder. */
+	cmd_handler_t func;
+};
+
+/**@brief Event decoder structure. */
+struct rp_ser_evt {
+	/** Event code. */
+	uint8_t evt;
+
+	/** Event decoder. */
+	evt_handler_t func;
+};
+
+/**@brief Commands and events decoder structure. */
+struct rp_ser_decoders {
+	/** Command section start address. */
+	const struct rp_ser_cmd *cmd_begin;
+
+	/** Command section end address. */
+	const struct rp_ser_cmd *cmd_end;
+
+	/** Event section start address. */
+	const struct rp_ser_evt *evt_begin;
+
+	/** Event section end address. */
+	const struct rp_ser_evt *evt_end;
+};
+
+/**@brief Helper macro for creating command decoder. All comands decoders have to be assigned
+ *        to proper Remote Procedure Serialization instance. After receiving a command the
+ *        command decoder function is searching in command data memory section based on command
+ *        number and called.
+ *
+ * @param[in] _rp_inst Remote Procedure Serialization instance.
+ * @param[in] _name Command decoder name.
+ * @param[in] _cmd Command number.
+ * @param[in] _handler Command decoder function @ref cmd_handler_t.
+ */
+#define RP_SER_CMD_DECODER(_rp_inst, _name, _cmd, _handler)					 \
+	RP_STATIC_ASSERT(&_rp_inst != NULL, "Invalid instance");                                 \
+	RP_STATIC_ASSERT(_cmd <= 0xFF, "Command out of range");                                  \
+	const struct rp_ser_cmd RP_CONCAT(_name, _cmd_dec) __used				 \
+	__attribute__((__section__(RP_STRINGIFY(RP_CONCAT(rp_ser_cmd_decoder_, _rp_inst))))) = { \
```

Comment:
It would be more comfortable to have this in linker script but it is no needed. This section can be defined here, in case in linker we would need to modify linker script in our Zephyr fork as I know which doesn't seem like a good solution but I might be wrong

Reply:
You can use `zephyr_linker_sources` and have the linker script additions in this folder.

Reply:
__attribute__((__section__() )) does not define a section. It only places the struct in a section.

Where is the section that the struct is placed in defined?

Could you perhaps upload a map file, and the preprocessed LD script, that would show how this struct is placed.

Reply:
This is from map file:

```
rp_ser_cmd_decoder_entropy_ser
                0x0000000001005638       0x10
 rp_ser_cmd_decoder_entropy_ser
                0x0000000001005638       0x10 app/libapp.a(entropy_ser.c.obj)
                0x0000000001005638                entropy_get_cmd_dec
                0x0000000001005640                entropy_init_cmd_dec
                0x0000000001005660                . = ALIGN (_region_min_align)
                0x0000000001005660                _image_rom_end = .
                0x0000000021000000                . = 0x21000000
                0x0000000021000000                . = ALIGN (_region_min_align)
                0x0000000021000000                _image_ram_start = .
```

Reply:
section start and stop symbol are also generated

Reply:
Yes, it's placed, but you can't get the address and size of the section, which is often the point of placing things in sections.

Reply:
Could you upload the full map and preprocessed LD script file?

Reply:
You can get in, in this case you have two symbol:
__start_rp_ser_cmd_decoder_entropy_ser - section start address
__stop_rp_ser_cmd_decoder_entropy_ser - section end address

So you can iterate over this section and caculate it size.

Reply:
It might be that it is placed because we use the "--place-orphan-sections" argument to the linker. Which means that when the linker gets a section it doesn't recognize, it makes an educated guess and places it where it sees fit, but not necessarily where we want it to be.

I would also appreciate an explanation of why we want to place this symbol in this section.

Reply:
And access this by
`extern const struct rp_ser_decoders  __start_rp_ser_cmd_decoder_entropy_ser []`
`extern const struct rp_ser_decoders  __stop_rp_ser_cmd_decoder_entropy_ser []`

Reply:
I've never seen GCC generate start and end symbols before. Are you using GCC?

Reply:
I believe it does so when an orphan section is found and the linker is told to place it anyway.

Reply:
Yes GCC. It seems like event_menager in ncs uses section in the same way as this library. I want to place this symbol in section to have all decoder grouped per instance in one place and easily iterate to look for matching decoders. If it is possible to have section per instance, where section instatnce count is any I can use `zephyr_linker_sources ` but I have to put linker script in Zephyr port in nrf and when user would want to use bare-metal or different OS then they have to create linker script with this section.

Reply:
If I could get steps to build this library I could generate a map file and LD script myself.

branches, sample, board etc.

Reply:
I found docs for it: https://sourceware.org/binutils/docs-2.20/ld/Orphan-Sections.html

Reply:
Hi, so this is the situation. In zephyr/NCS we usually link with
-Wl,--orphan-handling=error, and for good reason, but this subsystem
is requiring that we link with -Wl,--orphan-handling=place.

The difference is what the linker does when it's told to link in a
symbol that has an input section that it doesn't recognize. When there are
no matches on any of the input sections in the LD script it can
either error-out, warn, or it can look at the symbol's attrubtes, like
if it's code or data, and make a best guess at placing it somewhere.

The problem with allowing the LD to guess where it should be put is
that the assumptions that we make when writing the LD script can be
corrupted. Take for instance this snippet of LD script:

```
	_image_rodata_end = .;
        . = ALIGN(mpu_align);
	_image_rom_end = .;
```

The writer and reader of this LD script might make the following
assumption:

> The end of rom (_image_rom_end) ends with filler such that it is
> aligned according to the MPU requirements.

This assumption is guaranteed to be true if and only if we use
--orphan-handling=error. When the LD is allowed to place orphan
sections it could place entropy_ser at any point in the above
snippet.

If entropy_ser was placed just before _image_rom_end the alignment
would be corrupted and the MPU would fail.

If entropy_ser was placed just before _image_rodata_end, then any MMU
code that tried to protect the rodata from being accidentially
executed would fail. Because entropy_ser contains code, not rodata.

Now, it just happens to be that the particular version of the
particular linker we used placed entropy_ser exactly just after
_image_rodata_end and therefore we have no issues right now.

But this is by happy accident is not guaranteed to be true for
future toolchain releases, future toolchain support, or future LD script
versions.

So to be future-proof, we need to use --orphan-handling=error and
write/generate some LD script that explicitly places it somewhere that
we know it is safe for it to be.

https://sourceware.org/binutils/docs/ld/Orphan-Sections.html

Reply:
That's very informative, thanks, @SebastianBoe!

Reply:
@SebastianBoe Thanks, I understand your explanation, I will consider writing the LD script, but I not sure if it is possible in easy way. Do you have any idea that it is possible to write this LD script in such a way that it will be able to generate symbol for each rp_ser instance(some kind of subsection in section ?) where user can define any count of them ? What exactly `zephyr_linker_sources` do, Do I have to create a whole linker script when I using it(it repleacing default script) ?

Reply:
> Do I have to create a whole linker script when I using it(it repleacing default script) ?

No it allows you to insert snippets into predefined locations in the existing script.

Take a look: https://github.com/NordicPlayground/fw-nrfconnect-zephyr/blob/master/cmake/extensions.cmake#L866